scripts/builtin/xgboost.dml

```diff
@@ -0,0 +1,780 @@
+        log_odds = 0
+      else
+        log_odds = log(odds / (1 - odds))
+      x = (log_odds + learning_rate * sum(output_values))
+      e = 2.7182818284
+      new_prediction[entry,] = e^x / (1 + e^x)
+    }
+  }
+}
+
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    row_vector: nx1 vector, one sample with n features
+# INPUT:    curr_M: The current M matrix with the current tree
+# INPUT:    current_node: the start node of the current tree
+# OUTPUT:   output_value: the prediction for the current sample at the current tree
+getOutputValueForEntry = function(Matrix[Double] row_vector, Matrix[Double] curr_M, Matrix[Double] current_node)
+    return (Double output_value) {
+    assert(nrow(row_vector) == 1)
+    assert(as.scalar(current_node[1,]) == 1)
+
+    cur_node_index = 1 # cant take the node id because its diff to the index
+
+    # iterate over one tree and find output value
+    while(as.scalar(current_node[5,]) != 0.0) # until leaf
+    {
+      used_feature = as.scalar(current_node[4,])
+      feature_is_scalar = as.scalar(current_node[5,]) == 1
+      if(feature_is_scalar) #SCALAR values
+      {
+        if(as.scalar(row_vector[,used_feature]) < as.scalar(current_node[6,])) # go left
+        {
+          cur_node_index = (cur_node_index + as.scalar(current_node[3,]))
+          current_node = curr_M[,cur_node_index]
+        }
+        else #  go right
+        {
+          cur_node_index = cur_node_index + as.scalar(current_node[3,]) + 1
+          current_node = curr_M[,cur_node_index]
+        }
+      }
+      else # CATEGORICAL values
+      {
+        assert(as.scalar(current_node[5,]) == 2)
+        if(as.scalar(row_vector[,used_feature]) == 1) # go left
+        {
+            cur_node_index = (cur_node_index + as.scalar(current_node[3,]))
+            current_node = curr_M[,cur_node_index]
+        }
+        else # go right
+        {
+            assert(as.scalar(row_vector[,used_feature]) == 0)
+            cur_node_index = cur_node_index + as.scalar(current_node[3,]) + 1
+            current_node = curr_M[,cur_node_index]
+        }
+      }
+    }
+    output_value = as.scalar(current_node[6,])
+}
+
+#-----------------------------------------------------------------------------------------------------------------------
+# INPUT:    X: nxn matrix, original input matrix
+# INPUT:    y: nx1 vector of my target values
+# INPUT:    R The matrix R which for each feature in X contains the following information
+#					- R[,2]: 1 (scalar feature)
+#				    - R[,1]: 2 (categorical feature)
+# INPUT:    M: the nxn output matrix
+# INPUT:    prediction: nx1 vector, my current predictions for my target value y
+# INPUT:    tree_id: The current tree id, starting at 1
+# OUTPUT:   M: the current M matrix of this tree
+buildOneTree = function(Matrix[Double] X, Matrix[Double] y, Matrix[Double] R, Integer sml_type, Integer max_depth,
```

Comment:
I would split this function into a Regression version and a Classification version, since in the innermost part you split based on this (sml_type).

Reply:
done in https://github.com/apache/systemds/pull/1334/commits/c750df72571574b0c59d2cec7a732d5d6a558ef2